### PR TITLE
chore: Remove attrdict from model hub

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Description

`attrdict` is broken since Python 3.9.  Possible replacements do not meet our dependency requirements. This PR implements our own `AttrDict` as a utility class.

Replaces https://github.com/determined-ai/determined/pull/8541

## Test Plan

Unit tests pass. Also try any `mmdetection` and `hugginface` example.

## Checklist

- [C] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
MLG-1442

### CODE REVIEW

File: model_hub/model_hub/mmdetection/_trial.py [hunk 1]
```diff
@@ -55,8 +54,8 @@ class MMDetTrial(det_torch.PyTorchTrial):
 
     def __init__(self, context: det_torch.PyTorchTrialContext) -> None:
         self.context = context
-        self.hparams = attrdict.AttrDict(context.get_hparams())
-        self.data_config = attrdict.AttrDict(context.get_data_config())
+        self.hparams = context.get_hparams()
+        self.data_config = context.get_data_config()
```
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply model_hub/model_hub/mmdetection/_trial.py:58]: In the PyTorchTrialContext, hparams is an `Optional[Dict]`, which is what we're passing here. Are users of MMDetTrial doing an `__init__`, then changing hparams after the fact? My assumption was that hparams are expected to be passed as a dict or read from a file as a dict and that's the end of it.

[the remark this answers is not in the document]
Comment (MikhailKardash) [reply model_hub/model_hub/mmdetection/_trial.py:58]: I can change this PR to change all dict objects to SimpleNamespace, but I'm not sure what the resulting benefit is.

File: model_hub/model_hub/utils.py [hunk 2]
```diff
@@ -86,3 +86,19 @@ def compute_num_training_steps(experiment_config: Dict, global_batch_size: int)
...
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        for key in self.keys():
+            if isinstance(self[key], dict):
+                self[key] = AttrDict(self[key])
+            super().__setattr__(key, self[key])
+
+    def __setattr__(self, name: str, val: Any) -> None:
+        super().__setattr__(name, val)
+        super().__setitem__(name, val)
```
Comment (rb-determined-ai) [inline model_hub/model_hub/utils.py:101]: This seems really fishy to me...

File: model_hub/model_hub/huggingface/_trial.py [hunk 3]
```diff
@@ -265,17 +266,17 @@ def __init__(self, context: det_torch.PyTorchTrialContext) -> None:
...
 
-        if "num_training_steps" not in self.hparams:
+        if not hasattr(self.hparams, "num_training_steps"):
             # Compute the total number of training iterations used to configure the
             # learning rate scheduler.
-            self.hparams.num_training_steps = model_hub.utils.compute_num_training_steps(
+            self.hparams.num_training_steps = utils.compute_num_training_steps(
                 self.context.get_experiment_config(), self.context.get_global_batch_size()
             )
-        if "use_pretrained_weights" not in self.hparams:
```
Comment (rb-determined-ai) [inline model_hub/model_hub/huggingface/_trial.py:278]: are these still necessary?

File: model_hub/model_hub/huggingface/_config_parser.py [hunk 4]
```diff
@@ -283,10 +283,10 @@ def parse_dict_to_dataclasses(
     Args:
         dataclass_types: dataclasses with expected attributes.
         args: arguments that will be parsed to each of the dataclass_types.
-        as_dict: if true will return dictionary instead of AttrDict
+        as_dict: if true will return AttrDict instead of dict
 
     Returns:
-        One dictionary for each dataclass with keys filled in from args if found.
+        One namespace for each dataclass with keys filled in from args if found.
```
Comment (rb-determined-ai) [inline model_hub/model_hub/huggingface/_config_parser.py:289]: you've got other bits that are carryover from earlier editions, please fix

File: model_hub/model_hub/mmdetection/_trial.py [hunk 5]
```diff
@@ -91,7 +91,7 @@ def __init__(self, context: det_torch.PyTorchTrialContext) -> None:
 
         self.clip_grads_fn = None
         if self.cfg.optimizer_config.grad_clip is not None:
-            self.clip_grads_fn = lambda x: torch.nn.utils.clip_grad_norm_(
+            self.clip_grads_fn = lambda x: torch.nn.mmdetutils.clip_grad_norm_(
```
Comment (rb-determined-ai) [inline model_hub/model_hub/mmdetection/_trial.py:94]: yo

File: model_hub/model_hub/mmdetection/_trial.py [hunk 6]
```diff
@@ -68,8 +68,8 @@ def __init__(self, context: det_torch.PyTorchTrialContext) -> None:
         self.model.init_weights()
 
         # If use_pretrained, try loading pretrained weights for the mmcv config if available.
-        if self.hparams.use_pretrained:
-            ckpt_path, ckpt = utils.get_pretrained_ckpt_path("/tmp", self.hparams.config_file)
+        if hasattr(self.hparams, "use_pretrained"):
```
Comment (rb-determined-ai) [inline model_hub/model_hub/mmdetection/_trial.py:71]: also this can be put back, right?

**Review by rb-determined-ai — APPROVED** [review]
Very nice.

Rephrase the PR comment before landing.  `attrdict` doesn't "depend on a personal fork of attrdict3", rather `attrdict` is broken and possible replacements do not meet our dependency requirements.
